### PR TITLE
Support for AMX instructions incl. in future Intel Xeon code-named Diamond Rapids

### PR DESCRIPTION
**Pull request by egeor**
This PR adds support of the new AMX instructions

### CODE REVIEW

**Comment by alheinecke** [comment]
Is the branching ready for merging?

**Comment by egeor** [comment]
I believe so

File: samples/xgemm/gemm_kernel.c [hunk 1]
```diff
@@ -2009,23 +2017,51 @@ void ref_fused_matmul( gemm_def* i_gemm_def_in, void* l_a, void* l_b, void* l_c_
...
-      libxsmm_convert_bf8_f32( (libxsmm_bfloat8*)l_c_gold, (float*)l_c_tmp, i_gemm_def->ldc*i_gemm_def->n );
-      ref_matmul( i_gemm_def, l_a, l_b, l_c_tmp );
-      apply_relu(i_gemm_def, l_c_tmp, ref_fusion_arguments->relu_bitmask, l_use_bitmask);
-      libxsmm_rne_convert_fp32_bf8( (float*)l_c_tmp, (libxsmm_bfloat8*)l_c_gold, i_gemm_def->ldc*i_gemm_def->n );
-      i_gemm_def->c_type = LIBXSMM_DATATYPE_BF8;
-      libxsmm_free(l_c_tmp);
+      const char* env_arch = getenv("LIBXSMM_TARGET");
+      const int is_env_DMR = (env_arch == NULL) ? 0 : ( env_arch == libxsmm_stristr(env_arch, "dmr"));
+      int arch_cpuid = libxsmm_cpuid(NULL);
+      if ((!is_env_DMR && arch_cpuid < LIBXSMM_X86_AVX512_DMR) || (i_gemm_def->vnni_a == 0)) {
```
Comment (alheinecke) [inline samples/xgemm/gemm_kernel.c:2023]: Shouldn't that test not just only for flat?

File: samples/xgemm/gemm_kernel.c [hunk 2]
```diff
@@ -2009,23 +2017,51 @@ void ref_fused_matmul( gemm_def* i_gemm_def_in, void* l_a, void* l_b, void* l_c_
...
-      libxsmm_convert_hf8_f32( (libxsmm_hfloat8*)l_c_gold, (float*)l_c_tmp, i_gemm_def->ldc*i_gemm_def->n );
-      ref_matmul( i_gemm_def, l_a, l_b, l_c_tmp );
-      apply_relu(i_gemm_def, l_c_tmp, ref_fusion_arguments->relu_bitmask, l_use_bitmask);
-      libxsmm_rne_convert_fp32_hf8( (float*)l_c_tmp, (libxsmm_hfloat8*)l_c_gold, i_gemm_def->ldc*i_gemm_def->n );
-      i_gemm_def->c_type = LIBXSMM_DATATYPE_HF8;
-      libxsmm_free(l_c_tmp);
+      const char* env_arch = getenv("LIBXSMM_TARGET");
+      const int is_env_DMR = (env_arch == NULL) ? 0 : ( env_arch == libxsmm_stristr(env_arch, "dmr"));
+      int arch_cpuid = libxsmm_cpuid(NULL);
+      if ((!is_env_DMR && arch_cpuid < LIBXSMM_X86_AVX512_DMR) || (i_gemm_def->vnni_a == 0)) {
```
Comment (alheinecke) [inline samples/xgemm/gemm_kernel.c:2046]: see above?

**Review by alheinecke — COMMENTED** [review]
Only a minor nit on the test driver, if we need to check for the platform, I feel just flat layout would be enough.

File: samples/xgemm/gemm_kernel.c (hunk 1 above)
Comment (egeor) [reply samples/xgemm/gemm_kernel.c:2023]: Not really, the dmr path is the only one that computes the relu after the downconvert... the rest archs use the f32 intermediate buffer...

File: samples/xgemm/gemm_kernel.c (hunk 2 above)
Comment (egeor) [reply samples/xgemm/gemm_kernel.c:2046]: same as above

File: samples/xgemm/gemm_kernel.c (hunk 1 above)
Comment (alheinecke) [reply samples/xgemm/gemm_kernel.c:2023]: ah I see so this is as otherwise correctness checks will fail

**Review by alheinecke — APPROVED** [review]
lgtm

**Pull request merged**